Debugging c workshop (#283)
NuevoFoundation/workshops: content/english/debugging/02-identify-the-problem1.md

There are two kinds of errors: compiler and runtime errors. A compiler error usually indicates a problem with your syntax. Perhaps you wanted to express an idea in the program but didn't adhere to the rules of the programming language. These are caught when you compile your program.

Compiler errors are nice because they are relatively easy to fix. In C, compilers like `gcc` and will usually give a lot of information about what went wrong during compiling. Let’s take a look at some of them. Open the **Shell** in the Repl.it below and run 

```bash
make CompilerErrors
```

Note that for this example, we will not be running the program - this program is meant to showcase some common compiler errors that you may encounter.

<iframe height="500px" width="100%" src="https://replit.com/@nuevofoundation/Debugging-Samples-C#CompilerErrors.c" scrolling="no" frameborder="no" allowtransparency="true" allowfullscreen="true" sandbox="allow-forms allow-pointer-lock allow-popups allow-same-origin allow-scripts allow-modals"></iframe>

The first error should look like this (or similar):

![Error 1: Missing Semicolon](../resources/w2-01.png)

> A compiler error contains the name of the file (`CompilerErrors.c`) followed by the line it is on (`12`) as well as where on that line(`19`). It then gives a brief name of the error - in this case, the compiler expected a semicolon `;` at the end of the line. The compiler also gives you the line of code and points to where your error actually occurred! You can simply add a semicolon to fix this error.

Reviewer: ```suggestion
A compiler error contains the name of the file (`CompilerErrors.c`) followed by the line it is on (`12`) as well as where on that line (`19`). It then gives a brief name of the error - in this case, the compiler expected a semicolon `;` at the end of the line. The compiler also gives you the line of code and points to where your error actually occurred! You can simply add a semicolon to fix this error.
```

The other error is complaining that there is an extra `}` when it didn't expect one. The fix for this one is also relatively simple: you can add the corresponding curly brace `{` next to `func()`.

{{% notice warning %}}

> A missing curly brace is often the reason for seemingly extraordinary errors. The compiler is not great at picking up where a curly brace should be, so if you're including a header file with a missing curly brace, it'll blow up in your face. In almost all cases, when the compiler outputs a number errors in code you didn't write, chances are it's a curly brace missing somewhere in your (or perhaps even the libary's) code!

Reviewer: ```suggestion
A missing curly brace is often the reason for seemingly extraordinary errors. The compiler is not great at picking up where a curly brace should be, so if you're including a header file with a missing curly brace, it'll blow up in your face. In almost all cases, when the compiler outputs a number of errors in code you didn't write, chances are it's a curly brace missing somewhere in your (or perhaps even the library's) code!
```

{{% /notice %}}

Once you've fixed it, you can compile the code again using the same command.

```bash
make CompilerErrors
```

Wait, there are more errors! As mentioned in the warning box, the compiler isn't great at picking up errors if there's a missing curly brace. A compiler error can "hide" others.

We'll use a table to showcase common compiler errors and their general cause.

| Error | Cause | General Fix |
|---|---|---|
| Missing semicolon | Missing semicolon. | Add the semicolon `;`.|
| Expected function body after function declarator | Missing curly bracket `{`, usually at the beginning of a function body declaration. | Add the curly brace `{` where it belongs. |
| Expected identifier or '('| In the context of a curly brace, usually means you have an extra dangling curly brace `{`. | Remove the curly brace or add a corresponding `}` after it.|
| Redefinition of ... | Somewhere in your code you declared a variable. Later you declared it again. | Rename the variables, or remove one of them. |

> | Use of undeclared type... | The compiler can't find the declaration of a type you want to use. | Usually this comes from improper `#include` directives (as types are usually declared in header files). Make sure there's no typos. |

Reviewer: ```suggestion
| Use of undeclared type... | The compiler can't find the declaration of a type you want to use. | Usually this comes from improper `#include` directives (as types are usually declared in header files). Make sure there are no typos. |
```

> |Must use `struct` (or `enum`) tag | C requires you to use `struct NAME_OF_TYPE` or `enum NAME_OF_ENUM` whenever you want to refer to the struct/enum type. | Usually C developers use a `typedef` statement so they don't have to type `struct` or `enum` to refer to those types. You can add those keywords yourself, though.|

Reviewer: ```suggestion
| Must use `struct` (or `enum`) tag | C requires you to use `struct NAME_OF_TYPE` or `enum NAME_OF_ENUM` whenever you want to refer to the struct/enum type. | Usually C developers use a `typedef` statement so they don't have to type `struct` or `enum` to refer to those types. You can add those keywords yourself, though. |
```

> |No member named... | In your struct, there's no field of the name you requested. | Probably a typo, or you haven't defined a field in your struct yet. |

Reviewer: ```suggestion
| No member named... | In your `struct`, there's no field of the name you requested. | Probably a typo, or you haven't defined a field in your `struct` yet. |

```
